datawig/hpo.py

```diff
@@ -0,0 +1,400 @@
+        
+        :param train_df: training data as dataframe
+        :param test_df: test data as dataframe; if not provided, a ratio of test_split of the
+                          training data are used as test data
+        :param hps: nested dictionary where hps[global][parameter_name] is list of parameters. Similarly,
+                          hps[column_name][parameter_name] is a list of parameter values for each input column.
+                          Further, hps[column_name]['type'] is in ['numeric', 'categorical', 'string'] and is
+                          inferred if not provided. See init method of HPO for further details.
+
+        :param strategy: 'random' for random search or 'grid' for exhaustive search
+        :param num_evals: number of evaluations for random search
+        :param max_running_hours: Time before the hpo run is terminated in hours
+        :param user_defined_scores: list with entries (Callable, str), where callable is a function
+                          accepting **kwargs true, predicted, confidence. Allows custom scoring functions.
+        :param hpo_run_name: Optional string identifier for this run.
+                          Allows to sequentially run hpo jobs and keep previous iterations
+        :param simple_imputer: SimpleImputer instance from which to inherit column names etc.
+
+        :return: None
+        """
+
+        self.output_path = simple_imputer.output_path
+
+        if user_defined_scores is None:
+            user_defined_scores = []
+
+        self.hps = hps
+        simple_imputer.check_data_types(train_df)  # infer data types, saved self.string_columns, self.numeric_columns
+
+        # process_hp_configurations(hps) uses self.hps to populate self.hps_flat
+        hps_flat = self.__preprocess_hps(train_df, simple_imputer)
+
+        # train/test split if no test data given
+        if test_df is None:
+            train_df, test_df = random_split(train_df, [.8, .2])
+
+        # sample configurations for random search
+        if strategy == 'random':
+            hps_flat = hps_flat.sample(n=min([num_evals, hps_flat.shape[0]]), random_state=10)
+
+        logger.info("Training starts for " + str(hps_flat.shape[0]) + "hyperparameter configurations.")
+
+        # iterate over hp configurations and fit models. This loop could be parallelized
+        start_time = time.time()
+        elapsed_time = 0
+
+        for hp_idx, (_, hp) in enumerate(hps_flat.iterrows()):
+            if elapsed_time > max_running_hours:
+                logger.info('Finishing hpo because max running time was reached.')
+                break
+
+            # if concat_columns True, set all n.a. hps to n.a. TODO
+            logger.info("Fitting hpo iteration " + str(hp_idx) + " with parameters\n\t" +
+                        '\n\t'.join([str(i) + ': ' + str(j) for i, j in hp.items()]))
+            name = hpo_run_name + str(hp_idx)
```

Comment:
you are reusing how the name is constructed in `simple_imputer.load_hpo_model`. maybe it is better to create the name in a function here and call it function also from `simple_imputer.load_hpo_model`.

Reply:
In `simple_imputer.load_hpo_model`, the name is read directly from the hpo_results table and not constructed in the same way as in `hpo.tune()`, where the user defined name is combined with an index. I agree though, that the construction of the name is not very elegant.